test/e2e/tests/chaos.go

```diff
@@ -0,0 +1,297 @@
+	// Block until disruptions is done
+	<-done
+	By("Wait for the apiserver to come back")
+	err := wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
+		healthStatus := 0
+		t.migrationClient.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+		if healthStatus != http.StatusOK {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		util.Failf("timed out waiting for apiserver to come back: %v", err)
+	}
+
+	// Verify the migrations can complete
+	By("Wait for the storage state of the CRD to change")
+	util.Logf("v1Hash is %s", v1Hash)
+	util.Logf("v2Hash is %s", v2Hash)
+
+	// Wait for discoveryPeriod + 1 minute to give the triggering controller enough time to detect and react.
+	err = wait.PollImmediate(10*time.Second, discoveryPeriod+1*time.Minute, func() (bool, error) {
+		crdStorageState, err := t.migrationClient.MigrationV1alpha1().StorageStates().Get("tests.migrationtest.k8s.io", metav1.GetOptions{})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+		if crdStorageState.Status.CurrentStorageVersionHash != v2Hash {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		util.Failf("%v", err)
+	}
+
+	By("Wait for all storage states to converge")
+	err = wait.PollImmediate(30*time.Second, 10*time.Minute, func() (bool, error) {
+		l, err := t.migrationClient.MigrationV1alpha1().StorageStates().List(metav1.ListOptions{})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+		for _, ss := range l.Items {
+			if len(ss.Status.PersistedStorageVersionHashes) == 1 && ss.Status.PersistedStorageVersionHashes[0] == ss.Status.CurrentStorageVersionHash {
+				continue
+			}
+			util.Logf("resource %v has persisted hashes %v, and current hash %s",
+				ss.Spec.Resource,
+				ss.Status.PersistedStorageVersionHashes,
+				ss.Status.CurrentStorageVersionHash)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		util.Failf("%v", err)
+	}
+
+	By("Migrations should have all completed")
+	l, err := t.migrationClient.MigrationV1alpha1().StorageVersionMigrations(namespaceName).List(metav1.ListOptions{})
+	if err != nil {
+		util.Failf("%v", err)
+	}
+	for _, m := range l.Items {
+		if !succeeded(m.Status.Conditions) {
+			util.Failf("unexpected in progress migration for resource %v", m.Spec.Resource)
+		}
+	}
+	// TODO: actually verify the etcd contents:
+	// kubectl log into the etcd pod
+	// etcd pod name: etcd-server-bootstrap-e2e-master
+	// namespace: kube-system
+}
+
+func (t *StorageMigratorChaosTest) Run(sem *chaosmonkey.Semaphore) {
+	var once sync.Once
+	ready := func() {
+		once.Do(func() {
+			sem.Ready()
+		})
+	}
+	defer ready()
+
+	t.Setup()
+	ready()
+	t.Test(sem.StopCh)
+}
+
+func apiserverRestartsFunc() {
+	args := []string{"compute", "--project", os.Getenv("PROJECT"), "ssh", "--zone", os.Getenv("KUBE_GCE_ZONE"), os.Getenv("CLUSTER_NAME") + "-master", "--command", "sudo pkill -9 kube-apiserver"}
```

Comment:
kubelet will quickly restart the api server after it's pkill'ed, what about moving the apiserver manifest to somewhere, wait some period of time, then move it back? this way you can test your code works even after some period of apiserver downtime

Reply:
Kubelet restarts the pod after an exponentially time off.